Documentation for "Immutable Secrets/ConfigMaps" feature (#19297)
kubernetes/website: content/en/docs/concepts/configuration/secret.md

Secret updates.
{{< /note >}}

{{< feature-state for_k8s_version="v1.18" state="alpha" >}}

For clusters that are extensively using secrets (having at least tens of thousands of unique

> secret to pod mounts), we designed the `Immutable Secrets/ConfigMaps` feature. It allows you

Author: done

to mark individual Secrets/ConfigMaps as `Immutable`, which blocks any changes to their data
and in return:
- protects you from accidental bad updates that could cause applications outages
- improves performance of your cluster by significantly reducing load on kube-apiserver, by
  closing watches for secrets marked as immutable

> 

Author: Added (btw - it's not possible to change the data once mark as immutable - even by priviliged user).

> 

Reviewer: > it's not possible to change the data once mark as immutable - even by priviliged user

If I really want to make that change, it sounds like I'd need to delete the cluster and recreate it, then? Or can I delete the Secret and recreate it?

> 

Reviewer: :thought_balloon: are there any security implications, for example if a user is able to mark a ConfigMap immutable and thereby break a process that expects to be able to modify it?

> 

Author: > Or can I delete the Secret and recreate it?

You can delete the secret and create a new one, but that won't be picked by pods that are mounting it - you will also need to recreate those pods.

> are there any security implications, for example if a user is able to mark a ConfigMap immutable and thereby break a process that expects to be able to modify it?

If a user is able to modify a config map that automation should only modify, it can already break stuff, but setting malicious data in it.

> 

Reviewer: > You can delete the secret and create a new one, but that won't be picked by pods that are mounting it - you will also need to recreate those pods.

That's the sort of detail I think readers would be interested in.

> 

Author: Added - thanks!

Secret updates.
{{< /note >}}

{{< feature-state for_k8s_version="v1.18" state="alpha" >}}

For clusters that are extensively using secrets (having at least tens of thousands of unique
Secret to Pod mounts), Kubernetes has an _Immutable Secrets/ConfigMaps_ feature. It allows you
to mark individual Secrets/ConfigMaps as `Immutable`, which blocks any changes to their data
and in return:
- protects you from accidental bad updates that could cause applications outages

> - improves performance of your cluster by significantly reducing load on kube-apiserver, by

Reviewer 2: Hello @wojtek-t ,
I wonder if a new commit would fix the deploy preview. Could you add a new commit?
line 681, Possible edit: `- protects you from accidental (or unwanted) updates
line 699, spelling, secret

/cc @VineethReddy02

Reviewer 2: @wojtek-t , Would you rebase this PR? I am interested in viewing a preview of the page.
You may have picked up a previous version of the `zh` deployment.md file. The build error has been corrected.

Author: done

  closing watches for secrets marked as immutable

To use it, you need to enable `ImmutableEmphemeralVolumes`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) and then

> mark your Secret/ConfigMap by setting it `Immutable` field to true, e.g.:

Reviewer: nits:
```suggestion
mark your Secret/ConfigMap by setting its `Immutable` field to true; for example:
```

Author: done

```yaml
apiVersion: v1
kind: Secret
metadata:
  ...
data:
  ...
immutable: true
```

Once marked as immutable, it is NOT possible to revert this change nor to mutate the contents

> of data field. You can only delete and recreate the secrete, but that will not be picked by

Reviewer: nit:
```suggestion
of data field. You can only delete and recreate the Secret, but that will not be picked by
```

Also, I'd put this in a `{{< note >}}` or `{{< caution >}}` block to highlight the gotcha about Pod replacement.

Author: done

Secret updates.
{{< /note >}}

{{< feature-state for_k8s_version="v1.18" state="alpha" >}}

For clusters that are extensively using secrets (having at least tens of thousands of unique
Secret to Pod mounts), Kubernetes has an _Immutable Secrets/ConfigMaps_ feature. It allows you
to mark individual Secrets/ConfigMaps as `Immutable`, which blocks any changes to their data

> and in return:

Reviewer: Can I use an immutable Secret if I only have a few hundred Secrets in my cluster?

Author: Yes - but I don't see where the text suggests you can't.

Reviewer: > For clusters that are extensively using secrets (having at least tens of thousands of unique Secret to Pod mounts), Kubernetes has an _Immutable Secrets/ConfigMaps_ feature.

This could read like a requirement.

Author: What do you suggest? [I don't read it as a requirement]

- protects you from accidental (or unwanted) updates that could cause applications outages
- improves performance of your cluster by significantly reducing load on kube-apiserver, by
  closing watches for secrets marked as immutable

To use it, you need to enable `ImmutableEmphemeralVolumes`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) and then

> mark your Secret/ConfigMap by setting its `Immutable` field to true, e.g.:

Reviewer: nit:
```suggestion
mark your Secret or ConfigMap, by setting its `immutable` field to `true`;
for example:
```

Author: done

Secret updates.
{{< /note >}}

> {{< feature-state for_k8s_version="v1.18" state="alpha" >}}

Reviewer 2: @wojtek-t , Here are some suggestions.
Could you create a sub-heading to introduce this section?
Right now this content is under `Using Secrets` and `mounted-secrets-are-updated-automatically`.
Not sure if this makes sense?
````
#### Immutable Secrets and ConfigMaps
````

Possible edit to introduce feature:

The Kubernetes alpha feature _Immutable Secrets and ConfigMaps_ provides an option to set individual Secrets and ConfigMaps as immutable. For clusters that extensively use secrets (at least tens of thousands of unique Secret to Pod mounts), preventing changes to this data has the following advantages:
<!-- new line here, then list the advantages -->

For clusters that are extensively using secrets (having at least tens of thousands of unique
Secret to Pod mounts), Kubernetes has an _Immutable Secrets/ConfigMaps_ feature. It allows you
to mark individual Secrets/ConfigMaps as `Immutable`, which blocks any changes to their data
and in return:
- protects you from accidental (or unwanted) updates that could cause applications outages
- improves performance of your cluster by significantly reducing load on kube-apiserver, by

> closing watches for secrets marked as immutable.

Reviewer 2: nit, Line 686, `closing watches`?
- improves performance of your cluster by significantly reducing load on the kube-apiserver and closing watches for secrets marked as immutable.

Author: I think that "by" is better - because the performance is improved by the fact that we're closing those watches.

>   closing watches for secrets marked as immutable

Reviewer: ```suggestion
closing watches for secrets marked as immutable.
```

Author: done

> To use it, you need to enable `ImmutableEmphemeralVolumes`

Reviewer 2: Possible edit, something like:

To use this feature, enable the [ImmutableEmphemeralVolumes feature gate](...)
and set your Secret or ConfigMap `immutable` field to `true`. For example:

Author: done

[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) and then
mark your Secret or ConfigMap, by setting its `immutable` field to `true`, e.g.:
```yaml
apiVersion: v1
kind: Secret
metadata:
  ...
data:
  ...
immutable: true
```

{{< note >}}
Once marked as immutable, it is NOT possible to revert this change nor to mutate the contents

> of data field. You can only delete and recreate the Secret, but that will not be picked by

Reviewer 2: Possible edit, pick and choose what makes sense:
Once a Secret is marked as immutable, it is *not* possible to revert this change nor to mutate the contents of the `data` field. You can only delete and recreate the Secret. Existing Pods maintain a mount point to the deleted Secret. It is recommended to recreate these Pods.

Author: done

Secret updates.
{{< /note >}}

{{< feature-state for_k8s_version="v1.18" state="alpha" >}}

The Kubernetes alpha feature _Immutable Secrets and ConfigMaps_ provides an option to set
individual Secrets and ConfigMaps as immutable. For clusters that extensively use Secrets
(at least tens of thousands of unique Secret to Pod mounts), preventing changes to their
data has the following advantages:

> - protects you from accidental (or unwanted) updates that could cause applications outages

Reviewer 2: @wojtek-t , Changes look good. One issue:
Add a newline at line 685 before the first list item (the list is not rendering)
Thanks!
